lib/api/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs

```diff
@@ -99,3 +172,322 @@ instance FromJSON TxMetadataWithSchema where
+  where
+    conv :: Aeson.Value -> Either TxMetadataJsonSchemaError TxMetadataValue
+    conv Aeson.Null   = Left TxMetadataJsonNullNotAllowed
+    conv Aeson.Bool{} = Left TxMetadataJsonBoolNotAllowed
+
+    conv (Aeson.Number d) =
+      case Scientific.floatingOrInteger d :: Either Double Integer of
+        Left  n -> Left (TxMetadataJsonNumberNotInteger n)
+        Right n -> Right (TxMetaNumber n)
+
+    conv (Aeson.String s)
+      | Just s' <- T.stripPrefix bytesPrefix s
+      , let bs' = T.encodeUtf8 s'
+      , Right bs <- B16.decode bs'
+      , not (B8.any (\c -> c >= 'A' && c <= 'F') bs')
+      = Right (TxMetaBytes bs)
+
+    conv (Aeson.String s) = Right (TxMetaText s)
+
+    conv (Aeson.Array vs) =
+        fmap TxMetaList
+      . traverse conv
+      $ V.toList vs
+
+    conv (Aeson.Object kvs) =
+        fmap
+        ( TxMetaMap
+        . sortCanonicalForCbor
+        )
+      . traverse (\(k,v) -> (,) (convKey k) <$> conv v)
+      . fmap (first Aeson.toText)
+      $ Aeson.toList kvs
+
+    convKey :: Text -> TxMetadataValue
+    convKey s =
+      fromMaybe (TxMetaText s) $
+      parseAll ((fmap TxMetaNumber pSigned <* Atto.endOfInput)
+            <|> (fmap TxMetaBytes  pBytes  <* Atto.endOfInput)) s
+
+bytesPrefix :: Text
+bytesPrefix = "0x"
+
+parseAll :: Atto.Parser a -> Text -> Maybe a
+parseAll p =
+    either (const Nothing) Just
+    . Atto.parseOnly p
+    . T.encodeUtf8
+
+pUnsigned :: Atto.Parser Integer
+pUnsigned = do
+    bs <- Atto.takeWhile1 Atto.isDigit
+    -- no redundant leading 0s allowed, or we cannot round-trip properly
+    guard (not (BS.length bs > 1 && B8.head bs == '0'))
+    return $! BS.foldl' step 0 bs
+  where
+    step a w = a * 10 + fromIntegral (w - 48)
+
+pSigned :: Atto.Parser Integer
+pSigned = Atto.signed pUnsigned
+
+pBytes :: Atto.Parser ByteString
+pBytes = do
+  _ <- Atto.string "0x"
+  remaining <- Atto.takeByteString
+  when (B8.any hexUpper remaining) $ fail ("Unexpected uppercase hex characters in " <> show remaining)
+  case B16.decode remaining of
+    Right bs -> return bs
+    _ -> fail ("Expecting base16 encoded string, found: " <> show remaining)
+  where
+    hexUpper c = c >= 'A' && c <= 'F'
+
+sortCanonicalForCbor
+    :: [(TxMetadataValue, TxMetadataValue)] -> [(TxMetadataValue, TxMetadataValue)]
+sortCanonicalForCbor =
+  map snd
+  . L.sortOn fst
+  . map (\e@(k, _) -> (CBOR.uintegerFromBytes $ serialiseKey k, e))
+    where
+      serialiseKey = CBOR.serialize' CBOR.shelleyProtVer . toShelleyMetadatum
+
+toShelleyMetadatum :: TxMetadataValue -> Ledger.Metadatum
+toShelleyMetadatum (TxMetaNumber x) = Ledger.I x
+toShelleyMetadatum (TxMetaBytes  x) = Ledger.B x
+toShelleyMetadatum (TxMetaText   x) = Ledger.S x
+toShelleyMetadatum (TxMetaList  xs) =
+    Ledger.List [ toShelleyMetadatum x | x <- xs ]
+toShelleyMetadatum (TxMetaMap   xs) =
+    Ledger.Map [ (toShelleyMetadatum k,
+                      toShelleyMetadatum v)
+                | (k,v) <- xs ]
+
```

Comment:
I insist: Everything below this point should go into a module

`Cardano.Wallet.Primitive.Types.Tx.TxMetadata.Encryption`

in the `/primitive` directory. 🤓 By adjusting the types as I suggest, this will become possible. This work is necessary to achieve a good separation of concerns.